iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/SnapshotDiffApplier.java

```diff
@@ -0,0 +1,422 @@
+        Map<String, SnapshotRef> existingRefs) {
+
+      // Compute all index maps once
+      Map<Long, Snapshot> providedSnapshotByIds =
+          providedSnapshots.stream()
+              .collect(
+                  Collectors.toMap(
+                      Snapshot::snapshotId, s -> s, (existing, replacement) -> existing));
+      Map<Long, Snapshot> existingSnapshotByIds =
+          existingSnapshots.stream()
+              .collect(
+                  Collectors.toMap(
+                      Snapshot::snapshotId, s -> s, (existing, replacement) -> existing));
+      Set<Long> existingBranchRefIds =
+          existingRefs.values().stream().map(SnapshotRef::snapshotId).collect(Collectors.toSet());
+      Set<Long> providedBranchRefIds =
+          providedRefs.values().stream().map(SnapshotRef::snapshotId).collect(Collectors.toSet());
+
+      // Compute changes
+      List<Snapshot> newSnapshots =
+          providedSnapshots.stream()
+              .filter(s -> !existingSnapshotByIds.containsKey(s.snapshotId()))
+              .collect(Collectors.toList());
+      List<Snapshot> deletedSnapshots =
+          existingSnapshots.stream()
+              .filter(s -> !providedSnapshotByIds.containsKey(s.snapshotId()))
+              .collect(Collectors.toList());
+
+      // Categorize snapshots
+      List<Snapshot> stagedSnapshots =
+          newSnapshots.stream()
+              .filter(s -> s.summary().containsKey(SnapshotSummary.STAGED_WAP_ID_PROP))
+              .collect(Collectors.toList());
+
+      // Compute source IDs for cherry-pick operations
+      Set<Long> cherryPickSourceIds =
+          providedSnapshots.stream()
+              .filter(
+                  s ->
+                      s.summary() != null
+                          && s.summary().containsKey(SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP))
+              .map(s -> Long.parseLong(s.summary().get(SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP)))
+              .collect(Collectors.toSet());
+
+      List<Snapshot> cherryPickedSnapshots =
+          providedSnapshots.stream()
+              .filter(
+                  provided -> {
+                    // Only consider EXISTING snapshots as cherry-picked
+                    Snapshot existing = existingSnapshotByIds.get(provided.snapshotId());
+                    if (existing == null) {
+                      return false;
+                    }
+
+                    // Is source of cherry-pick
+                    if (cherryPickSourceIds.contains(provided.snapshotId())) {
+                      return true;
+                    }
+
+                    // WAP snapshot being published (staged → branch transition)
+                    boolean hasWapId =
+                        provided.summary() != null
+                            && provided.summary().containsKey(SnapshotSummary.STAGED_WAP_ID_PROP);
+                    boolean wasStaged = !existingBranchRefIds.contains(provided.snapshotId());
```

Comment:
```suggestion
                    // TODO: This works for MAIN branch only,  but fails in the branch scenario and should be revisited in followup PR 
                    // Snapshot exists on branch-A
                    // Cherry-pick to branch-B
                    // Would be classified as NOT wasStaged (because it's in existingBranchRefIds)
                    // Wouldn't be detected as cherry-picked
                    boolean wasStaged = !existingBranchRefIds.contains(provided.snapshotId());
```